### DISCUSSION TITLE
Port Forwarding returns 502 Bad Gateway

### DISCUSSION THREAD

**Discussion by hashfx** [community/community#28563] [Codespaces]
When running a react app on port 3000, and Express API on port 5000 gives 502 Bad Gateway Error.

![Screenshot](https://user-images.githubusercontent.com/69109482/183244840-5101006d-e655-4d67-b670-fe3bce0b505e.png)

**Comment by augustoaraujoo** [comment]
try running this command in your terminal, here is how to run it, killall node --> https://stackoverflow.com/questions/14790910/stop-all-instances-of-node-js-server,

then try to activate the app

**Reply by hashfx** [reply]
Thanks for your reply.
I killed the whole setup and rebuild it, still it didn't work. The server won't work in between (shows 502 error) but would work after sometime (no idea why this happens). May be it's from "their" side👀

**Comment by jaclynj** [comment]
Is this on web or the vscode desktop extension? 

Are you seeing an error in your server logs?

**Reply by hashfx** [reply]
This is on web. There's no error in server log, it's directly rendered on the screen.

**Comment by legomushroom** [comment]
@hashfx did you try to switch port protocol(HTTP vs HTTPS) in the ports view in the VSCode? Sometime the service fails to detect a correct protocol, so switching to a correct one manually might help. You can `right-click on the port` -> `Change Port Protocol` -> select `HTTP` or `HTTPS`.

**Reply by hashfx** [reply]
Yeah I did, this error is occasional. Server runs for sometime, gets 502 error, starts running again after sometime.

**Comment by nightpool** [comment]
I'm having the same issue on web. It seemed to happen after I rebuilt my container

**Comment by azzazkhan** [comment]
I am getting the same issue, using my custom-built template ([azzazkhan/dev-container](https://github.com/azzazkhan/dev-container)).

**Comment by jnewland** [comment]
I've had a support ticket open with GitHub since April regarding a similar issue: I'm able to consistently reproduce a 502 on Codespaces port forwarding by starting a Codespace that uses Docker Compose in its devcontainer configuration, forwarding a port, and then rebuilding the Codespace. This behavior is quite frustrating as it is often triggered while I'm iterating on my organization's Codespaces configuration, adding to the complexity of the already-time-consuming process of QAing these sorts of changes.

https://github.com/jnewland/devcontainer-example/tree/docker-compose-port-forward demonstrates the behavior, or did the last time I tested it.

GitHub support has advised me that a workaround for this problem is to stop forwarding the port *before rebuilding the Codespace*, which I haven't yet successfully validated as I only remember this bug when I experience the 502, not beforehand. Disabling port forwarding and re-enabling it doesn't seem to resolve the issue once you experience it: at this point I generally delete the Codespace and recreate it.

**Reply by j-abrams** [reply]
We have been attempting the suggested workaround to stop forwarding the port before rebuilding without success.  Any progress with regards to this problem?  Very frustrating and time-consuming deleting and recreating every time you need to launch the port!

**Reply by ankur0101** [reply]
I have started to observe this error at my codespace when I reach 75% of the free quota. I use python flask, generally after running command `flask --debug run`, it automatically starts forwarding port under ports tab. However in recent days, it has become very inconsistent :-(

**Reply by jnewland** [reply]
Support has let me know that this problem has been resolved.

**Reply by betaglot** [reply]
It has not been resolved!!

**Reply by mrhoodz** [reply]
Im also having this problem

**Reply by PaulDMendoza** [reply]
I think I'm also having this issue. I'm using docker compose.

**Comment by NidhoggDJoking** [comment]
I am getting the same issue!
It has not been resolved

<img src="https://user-images.githubusercontent.com/38490462/215432285-66072e33-3792-4315-9341-565f8bd042be.png" width="350px" />

**Comment by jetadidas** [comment]
Just happened to me right in the middle if a build!

**Comment by aitchjoe** [comment]
I had the same issue, but when I tried:

```
docker run -d -p 127.0.0.1:8080:80 docker.io/nginx
```

It is ok. So I used `ss -ltn` to check the difference and found that nginx bound to `127.0.0.1:8080` but my app bound to `[::1]:3000`, then I changed my webpack-dev-server config to `127.0.0.1:3000` and it is ok now.

**Comment by iqfareez** [comment]
Try starting the server on any IPv4: `0.0.0.0`. Example for Node JS based applications:

```
npm run start -- --host 0.0.0.0
```

For yarn:

```
yarn start --host 0.0.0.0
````

For angular, see this [answer](https://github.com/orgs/community/discussions/28563#discussioncomment-5224382). For vite, see [this](https://github.com/orgs/community/discussions/28563#discussioncomment-5438853).

**Reply by bradley-varol** [reply]
This works for me when running `vite dev` (`npm run dev -- --host 0.0.0.0`)

**Reply by somanath-goudar** [reply]
Thank You this worked for me

**Reply by KritagyaPrasad** [reply]
Thanks this worked for me as well.

**Reply by Mexmorize** [reply]
I tried this since it looked promising although my root directory doesn't have a package.json so I'm encountering an error. My root directly has multiple containers (in their own folders and some have their own package.json but others don't, like the database).

Any thoughts?

**Reply by mystique09** [reply]
this works for me, thanks

**Reply by ix-xerri** [reply]
Worked for me as well. Thank a lot!

**Reply by ht55ght55** [reply]
This is a common solution for cases where the server is not accessible outside of the local network or when trying to access it from another machine.

The reason for the 502 Bad Gateway Error when running a React app on port 3000 and an Express API on port 5000 is likely due to a misconfiguration in the reverse proxy server (e.g., Nginx, Apache) that is routing traffic between the two servers.

When the React app tries to make a request to the Express API on port 5000, the reverse proxy server is responsible for forwarding that request to the correct server. If the configuration is not set up correctly, the reverse proxy server may not be able to communicate with the Express API on port 5000, resulting in the 502 Bad Gateway Error.

Starting the server on IP address 0.0.0.0 may not directly solve the 502 Bad Gateway Error, but it can help ensure that the server is accessible from any IP address. To resolve the error, you will need to investigate the configuration of the reverse proxy server and ensure that it is correctly forwarding requests to the correct server and port.

**Reply by thyaravind** [reply]
I faced same issue when running a svelte app. This resolved it. Thanks

**Reply by ianwesterfield** [reply]
This did it, thanks @iqfareez !

I used `nx serve app --host 0.0.0.0`. It doesn't always give me the feel-good toast message or show the connected service, but the link works just the same. The underlying issue remains unknown, but that's GitHub's issue :)

**Reply by inegmdev** [reply]
@ianwesterfield @ht55ght55 
You're awsome, same goes for Vite, using this command `vite --host 0.0.0.0`.

**Reply by karneaud** [reply]
This did not work for me whne using `server.listen(2567,'0.0.0.0'....`

**Reply by adonyssantos** [reply]
> network

Thanks!! That worked for me.

**Reply by jhaby** [reply]
I was having this issue and it was killing my day and this soultion did it. Thanks

**Reply by wannysim** [reply]
Worked for me as well, in `yarn v3.5.0` ! Thanks 👍

**Reply by fe7ng654123** [reply]
> Try starting the server on any IPv4: `0.0.0.0`. Example for Node JS based applications:
> 
> ```
> npm run start -- --host 0.0.0.0
> ```
> 
> For yarn:
> 
> ```
> yarn start --host 0.0.0.0
> ```
> 
> For angular, see this [answer](https://github.com/orgs/community/discussions/28563#discussioncomment-5224382). For vite, see [this](https://github.com/orgs/community/discussions/28563#discussioncomment-5438853).

saved my day

**Reply by BentonKwok** [reply]
worked for me :)
 `npm run dev -- --host 0.0.0.0`

**Reply by SkylerMime** [reply]
This didn't fix the problem for me for my Vite app :/

**Comment by ghost** [comment]
Running `killall node` killed the terminal and the entire codespace for our team, not recommended

**Comment by tristan-f-r** [comment]
In general, if you're having this problem, host at `0.0.0.0`. Codespaces has trouble detecting localhost servers, unlike gitpod.

**Reply by Ni-011** [reply]
hey, i am working on a project that uses yarn dx to run a few scripts, it hosts it on 3000, how do i change it to 0.0.0.0 ?

**Comment by 0jdk** [comment]
is this resolved? im running aurelia app in codespace and getting this error
tried above workarounds, didn't help

**Comment by stevenlevey** [comment]
I have the exact same issue when trying to stand up a default sveltekit application. I've tried all the above suggestions and I get a 502 Gateway error every single time.

**Reply by tristan-f-r** [reply]
Odd, trying `--host` on my dev server usually worked for me.

**Reply by steven-levey-sp** [reply]
Ok I have an update. (Revisited Codespaces after moving to Codesandbox for a while)

If you run `npm run dev --host 0.0.0.0 --port 8080` that seems to fix it for me. So the key part was to add the --port 8080 as without that it just couldn't find it.

**Reply by SkylerMime** [reply]
This solution worked for me! Apparently Codespaces was having trouble with the default Vite port (5173), but 8080 works fine.

**Comment by Miki-Leo** [comment]
Hi, i have the same issue now... was someone able to fix it afterall?

**Comment by dkotama** [comment]
with angular , running `ng serve --host 0.0.0.0 ` fix the problem for me

**Comment by azeek21** [comment]
For anyone using `yarn` and having same problem. running `yarn start` gave me same bad gateway error.

Fix: `yarn ... --host 0.0.0.0` worked for me. NOTE: replace dots with the command you want. e.g: `dev`, `start` ...

**Comment by akshay-verma** [comment]
For anyone using `vite`, use `—host 0.0.0.0` explicitly in your commands.
Example - 
```
vite --port 5173 --host 0.0.0.0
vite preview --port 5555 --host 0.0.0.0
```

**Reply by mrhoodz** [reply]
why would someone pay for github code spaces if it cant do some a simple as showing you what you are making ??

**Comment by loghomefinishing** [comment]
"dev": "vite dev --host",
be sure to add the --host and it will work in package.json

**Reply by vskbot007** [reply]
it works . Thanks !!

**Reply by ValterJunior** [reply]
It worked like a charm on React + Vite!

**Comment by suggoooiii** [comment]
If you're using Vite you can change the following in your package.json 
`  "scripts": {
    "dev": "vite --host 0.0.0.0",
    "build": "vite build",
    "preview": "vite preview"
  },`

now you can run the following from your terminal `npm run dev`

**Comment by torodebout** [comment]
add `server.host = true` block in `vite.config.js`, will solve the issue , like this :
`export default defineConfig({
  plugins: [react()],
  server: {
    host: true,
  }
})`

**Comment by kioley** [comment]
Try changing both the **host** and the **port**, as it may not work on all ports
_example for **VITE**:_
`yarn vite --host 0.0.0.0 --port 3000`

**Comment by luke1212** [comment]
I am trying to develop a web application based on Angular in Codespace. When I did "npm run start", I got the local host URL in the terminal. After clicking on it I got a blank loading page which supposes to be the initial start page of angular if I run this same command locally.

**Reply by ianwesterfield** [reply]
Try ‘ng serve <your-app> —host=0.0.0.0’?

**Reply by luke1212** [reply]
Hi, ianwesterfield,

Thank you for your reply! I found a similar solution
I find this on another closed issue and it works! 
https://github.com/microsoft/vscode/issues/173432#issuecomment-1434885765

**Reply by ianwesterfield** [reply]
Cool beans - good luck!

**Comment by deepaksankar** [comment]
I just got this same error while using npm run dev for a simple react app I am creating

**Comment by thehackermonkey** [comment]
My vite project  worked after changing the port from `5173` to `3000`. My advice is to try port 3000

`package.json`

```
"scripts": {
    "dev": "vite dev --port 3000 --host",
...
}
```

**Comment by takumikuroyanagi** [comment]
Killing the running processes worked for my React Project.
Simply by restarting the codespaces.

![image](https://github.com/community/community/assets/133204210/97a13f92-8155-4678-88fd-5fce5f4fe577)

[Stopping and starting a codespace - GitHub Docs](https://docs.github.com/en/codespaces/developing-in-codespaces/stopping-and-starting-a-codespace)

`killall node` is not recommended.
> Running `killall node` killed the terminal and the entire codespace for our team, not recommended

**Comment by natecalc** [comment]
Hi all. 

I kept running into this 502 bad gateway error when i was in my `main` codespace, then i created a new branch. When i created the new branch the 502 bad gateway issue began happening.

The solve for this is to just stop the codespace, but restart it while you are on your newly created branch. Hope this helps.

**Comment by ghost** [comment]
In my case i get 502 bad gateway
at this URL `https://codespaceid-5173.preview.app.github.dev/`
but the app is working fine here `https://codespaceid-5173.app.github.dev/`

**Comment by howardwu1** [comment]
I also saw this issue with spring boot and port forwarding -- and this was also intermittent, especially when first creating the codespace (and even after multiple restarts from then).
